src/tmpfiles/tmpfiles.c

```diff
@@ -1973,6 +1985,44 @@ static int parse_line(const char *fname, unsigned line, const char *buffer) {
         if (!should_include_path(i.path))
                 return 0;
 
+        if (i.argument && IN_SET(i.type, COPY_FILES, CREATE_SYMLINK, CREATE_FILE, TRUNCATE_FILE, WRITE_FILE)) {
+                _cleanup_free_ char *resolved = NULL;
+
+                r = specifier_printf(i.argument, specifier_table, NULL, &resolved);
+                if (r == -ENOKEY) {
+                        log_notice("[%s:%u] Uninitialized /etc detected, skipping rule", fname, line);
+                        return 0;
+                }
+                if (r < 0) {
+                        const char *fmt;
+
+                        switch (i.type) {
+                        case COPY_FILES:
+                                fmt = "[%s:%u] Failed to substitute specifiers in copy source: %m";
+                                break;
+                        case CREATE_SYMLINK:
+                                fmt = "[%s:%u] Failed to substitute specifiers in symlink target: %m";
+                                break;
+                        case CREATE_FILE:
+                        case TRUNCATE_FILE:
+                        case WRITE_FILE:
+                                fmt = "[%s:%u] Failed to replace specifiers in parameter to write: %m";
+                                break;
+                        default:
+                                assert_not_reached("Unexpected command type.");
+                        }
+
+                        DISABLE_WARNING_FORMAT_NONLITERAL;
+                        log_error(fmt, fname, line);
+                        REENABLE_WARNING;
+                        return r;
```

Comment:
quite frankly, i don't really like that the specifier expansion is now partly done beforehand (as seen here) and partly at the moment of application (for the xattr). I'd prefer if we'd either do all expansion beforehand, or all at the moment of application, but not distribute this all over the place...

This means, either the xattr parsing needs to be moved somewhere early too, or the specifier expansion should stay where it is, but the errors from that need to be handled properly and propagated...

Reply:
well with these changes, the specifier expansion is self contained in parse_line() function at least. Before specifier_printf() was spread all over the place, so I would say that it's slightly better now.

That said we could simply create a function that deals with the specifier expansion and take care of all relevant command types. For RECURSIVE_SET_XATTR and SET_XATTR types, the function would do a special treatment.